README.md
# 云片 移动认证 Android SDK 接入指南

一键登录是云片提供的 APP 身份验证产品。整合三大运营商的手机号网关验证服务，可以替代 APP 上的注册、登录等场景需要使用到的短信验证码，优化 APP 用户在移动设备上的使用体验。

## 1. 整体集成流程

以下为一键登录的集成流程，整个集成流程是顺序进行的，在下一个步骤开始前请确保上一个步骤的操作都已正确完成。

#### 1.1. 获取应用 AppId

1、访问[云片官网](https://www.yunpian.com/entry?method=register)进行注册账号，联系客服或者销售申请开通移动认证服务。

2、成功开通服务后，进入[管理控制台](https://www.yunpian.com/admin/main)，进入移动认证的产品管理页面。

3、选择**新增应用**，填写应用名称以及 BundleId，系统会为该产品分配 **AppId**，应用进入审核状态，并联系客服进行审核，等待审核完成。

#### 2. 集成客户端 SDK

客户端 SDK 支持 **Android**、**iOS** 两大平台，最低支持的系统版本 A：Android 4.1、iOS 9.0 及以上，涉及到网站主的两个 API 请求以及与服务端的几个 API 请求。客户端 SDK 的业务使用流程为：

1、调用初始化接口，初始化一键登录 SDK。

2、在初始化接口的回调中，调用预取号接口。

3、调用一键登录接口，完成一键登录，将成功回调的 cid 返回给开发者自己的服务端。

####  3. 从接口获取校验结果(开发者服务端)

开发者服务端从客户端请求解析到相关参数后，接口能正确返回校验结果，即代表集成成功。

[云片移动认证服务端接入文档](https://github.com/yunpian/yunpian-onelogin-demo-android/blob/master/云片移动认证服务端接入文档.md)

## 开始使用

### 集成

需要确保主项目 build.gradle 文件中声明了 jcenter() 配置

```
implementation 'com.yunpian:onelogin:2.0.3'
```

#### 初始化

使用开发者自己的 appId 进行初始化 SDK

```java
QPOneLogin.getInstance().init(context, "your appId", callback);
```

#### 预取号

可以在取号之前预取号，避免取号时间太长

```java
QPOneLogin.getInstance().preGetToken(callback);
```

#### 取号

```java
QPOneLogin.getInstance().requestToken(themeConfig,new AbsQPResultCallback(){
		@Override
		public void onSuccess(String message) {

		}

		@Override
		public void onFail(String message) {

		}

		@Override
		public void onAuthActivityCreate(Activity activity) {
		    // 授权页面拉起回调
		}

		@Override
		public void onAuthWebActivityCreate(Activity activity) {
		    // 隐私条款页面拉起回调
		}

		@Override
		public void onLoginButtonClick() {
		    // 登录按钮点击回调
		}

		@Override
		public void onPrivacyCheckBoxClick(boolean isChecked) {
		    // CheckBox点击回调
		}

		@Override
		public void onPrivacyClick(String name, String url) {
		    // 隐私条款点击回调
		}
});
```

#### 关闭授权页面

```java
QPOneLogin.getInstance().dismissAuthActivity();
```

#### 打开短信回退界面

```java
QPOneLogin.getInstance().requestSmsToken(callback);
```

### 注册短信界面生命周期回调接口

```java
QPOneLogin.getInstance().registerSmsAuthActivityLifecycleCallback(lifecycleCallbacks);
```

### 自定义短信回退界面

覆盖相关布局文件或者资源文件即可。然后在短信界面的生命周期回调接口的 onActivityCreated  方法编写相关逻辑，例如

```java
QPOneLogin.getInstance().registerSmsAuthActivityLifecycleCallback(new SimpleActivityLifecycleCallback() {
            @Override
            public void onActivityCreated(final Activity activity, Bundle savedInstanceState) {
                LinearLayout rootLl = activity.findViewById(R.id.root);
                LinearLayout customView = (LinearLayout) getLayoutInflater().inflate(R.layout.custom_login, null);
                // 自定义逻辑
            }
        });
}
```

#### 设置日志输出

```java
QPOneLogin.getInstance().setLogEnable(enable);
```

#### 开发者自定义认证界面控件

在 requestToken() 方法之前调用。允许开发者在授权页面 titlebar 和 body 添加自定义的控件
注意：自定义的控件不允许覆盖 SDK 默认的 UI。

```java
QPOneLogin.getInstance().addOneLoginRegisterViewConfig(id, authRegisterViewConfig);
```

#### AuthRegisterViewConfig 接口

| 方法名 | 传参类型 | 说明 |
| --- | ------- | --- |
| setView | View | 开发者传入自定义的控件，需要设置好控件的布局属性，SDK 只支持 RelativeLayout 布局 |
| setRootViewId | int | 设置控件的位置。RootViewId.ROOT_VIEW_ID_TITLE_BAR 指标题栏，RootViewId.ROOT_VIEW_ID_BODY 指授权页空白处 |
| setCustomInterface | CustomInterface | 设置控件事件。CustomInterface是一个接口，实现了点击方法 |

#### 自定义 UI

设置背景图片：setAuthBGImgPath(String authBGImgPath)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| authBGImgPath | String | 设置背景图片。放在drawable文件下，以下背景图片路径与之保持一致 | gt_one_login_bg |

设置弹窗主题相关：setDialogTheme(boolean isDialogTheme, int dialogWidth, int dialogHeight, int dialogX, int dialogY, boolean isDialogBottom, boolean isWebViewDialogTheme)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| isDialogTheme | boolean | 是否使用弹窗模式，true为使用，false为不使用 | false |
| dialogWidth | int | 弹窗模式宽度，长宽的单位都为dp，以下单位与之保持一致 | 300 |
| dialogHeight | int | 弹窗模式高度 | 500 |
| dialogX | int | 授权⻚弹窗X偏移量（以屏幕⼼为原点） | 0 |
| dialogY | int | 授权⻚弹窗Y偏移量（以屏幕⼼为原点） | 0 |
| isDialogBottom | boolean | 授权⻚弹窗是否贴于屏幕底部true：显示到屏幕底部， dialogY参数设置将⽆效 false：不显示到屏幕底部，以 dialogY参数为准 | false |
| isWebViewDialogTheme | boolean | 隐私页面是否使用弹窗模式 | false |

设置状态栏颜色、字体颜色：setStatusBar(int statusBarColor,int navigationBarColor,boolean isLightColor)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| statusBarColor | int | 自定义状态栏背景颜色 | 0 |
| navigationBarColor | int | 自定义底部导航栏背景颜色 | 0 |
| isLightColor | boolean | 设置状态栏内容的颜色（只能黑白），true为黑色，false为白色 | false |

设置标题栏布局：setAuthNavLayout(int navColor, int authNavHeight, boolean authNavTransparent, boolean authNavGone)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| navColor | int | 标题栏颜色 | 0xFF3973FF |
| authNavHeight | int | 标题栏高度 | 49 |
| authNavTransparent | boolean | 标题栏是否透明 | true |
| authNavGone | boolean | 标题栏是否隐藏，此处的隐藏为View.GONE | false |

设置标题栏中间文字相关：setAuthNavTextView(String navText, int navTextColor, int navTextSize, boolean navTextNormal, String navWebViewText, int navWebViewTextColor, int navWebViewTextSize)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| navText | String | 文字设置 | 一键登录 |
| navTextColor | int | 字体颜色 | 0xFFFFFFFF |
| navTextSize | int | 字体大小，单位为sp,以下设置字体大小的单位与之保持一致 | 17 |
| navTextNormal | boolean | 设置是否隐私条款页面的标题栏中间文字使用默认值，true为使用navWebViewText，false为使用默认隐私条款的名字 | false |
| navWebViewText | String | 隐私条款页面的标题栏中间文字 | 服务条款 |
| navWebViewTextColor | int | 隐私条款页面的标题栏中间文字颜色 | 0xFF000000 |
| navWebViewTextSize | int | 隐私条款页面的标题栏中间文字大小 | 17 |

设置标题栏返回按钮相关：setAuthNavReturnImgView(String navReturnImgPath, int returnImgWidth, int returnImgHeight, boolean navReturnImgHidden, int returnImgOffsetX)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| navReturnImgPath | String | 文字设置 | gt_one_login_ic_chevron_left_black |
| returnImgWidth | int | 返回按钮图片宽度 | 24 |
| returnImgHeight | int | 返回按钮图片高度 | 24 |
| navReturnImgHidden | boolean | 返回按钮是否隐藏 | false |
| returnImgOffsetX | int | 返回按钮图片距离屏幕左边X轴偏移量 | 12 |

设置LOGO相关：setLogoImgView(String logoImgPath, int logoWidth, int logoHeight, boolean logoHidden, int logoOffsetY, int logoOffsetY_B, int logoOffsetX)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| logoImgPath | String | logo图片 | gt_one_login_logo |
| logoWidth | int | logo图片宽度 | 71 |
| logoHeight | int | logo图片高度 | 71 |
| logoHidden | boolean | logo是否隐藏 | false |
| logoOffsetY | int | logo相对于状态栏下边缘y偏移 | 125 |
| logoOffsetY_B | int | logo相对于底部y偏移 | 0 |
| logoOffsetX | int | logo相对于屏幕左边X轴偏移量,当为0时表示居中显示 | 0 |

设置号码相关：setNumberView(int numberColor, int numberSize, int numFieldOffsetY, int numFieldOffsetY_B, int numFieldOffsetX)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| numberColor | int | 号码栏字体颜色 | 0xFF3D424C |
| numberSize | int | 号码栏字体大小 | 24 |
| numFieldOffsetY | int | 号码栏相对于标题栏下边缘y偏移 | 200 |
| numFieldOffsetY_B | int | 号码栏相对于底部y偏移 | 0 |
| numFieldOffsetX | int | 号码栏相对于屏幕左边X轴偏移量,当为0时表示居中显示 | 0 |

设置号码栏字体相关：setNumberViewTypeface(Typeface numberViewTypeface)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| numberViewTypeface | Typeface | 号码栏的文字的字体 | Typeface.DEFAULT |

设置切换账号相关：setSwitchView(String switchText, int switchColor, int switchSize, boolean switchAccHidden, int switchAccOffsetY, int switchOffsetY_B, int switchOffsetX)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| switchText | String | 切换账号文字 | 切换账号 |
| switchColor | int | 切换账号字体颜色 | 0xFF3973FF |
| switchSize | int | 切换账号字体大小 | 14 |
| switchAccHidden | boolean | 切换账号是否隐藏 | false |
| switchAccOffsetY | int | 切换账号相对于标题栏下边缘y偏移 | 249 |
| switchOffsetY_B | int | 切换账号相对于底部y偏移 | 0 |
| switchOffsetX | int | 切换账号相对于屏幕左边X轴偏移量,当为0时表示居中显示 | 0 |

设置切换账号字体相关：setSwitchViewTypeface(Typeface switchViewTypeface)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| switchViewTypeface | Typeface | 切换账号的文字的字体 | Typeface.DEFAULT |

设置登录按钮布局：setLogBtnLayout(String loginImgPath, int logBtnWidth, int logBtnHeight, int logBtnOffsetY, int logBtnOffsetY_B, int logBtnOffsetX)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| loginImgPath | String | 登录按钮背景图片 | gt_one_login_btn_normal |
| logBtnWidth | int | 登录按钮宽度 | 268 |
| logBtnHeight | int | 登录按钮高度 | 36 |
| logBtnOffsetY | int | 登录按钮相对于标题栏下边缘y偏移 | 249 |
| logBtnOffsetY_B | int | 登录按钮相对于底部y偏移 | 0 |
| logBtnOffsetX | int | 登录按钮相对于屏幕左边X轴偏移量,当为0时表示居中显示 | 0 |

设置登录按钮中间文字相关：setLogBtnTextView(String loginButtonText, int loginButtonColor, int logBtnTextSize)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| loginButtonText | String | 文字设置 | 一键登录 |
| loginButtonColor | int | 文字颜色 | 0xFFFFFFFF |
| logBtnTextSize | int | 文字大小 | 15 |

设置登录按钮中间文字的字体相关：setLogBtnTextViewTypeface(Typeface logBtnTextViewTypeface)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| logBtnTextViewTypeface | Typeface | 登录按钮中间的文字的字体 | Typeface.DEFAULT |

设置loading图片相关：setLogBtnLoadingView(String loadingView, int loadingViewWidth, int loadingViewHeight, int loadingViewOffsetRight)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| loadingView | String | loading图片地址 | 一键登录 |
| loadingViewWidth | int | loading图片宽度 | 20 |
| loadingViewHeight | int | loading图片高度 | 20 |
| loadingViewOffsetRight | int | loading图片距离屏幕右边X轴偏移量 | 12 |

设置Slogan相关：setSloganView(int sloganColor, int sloganSize, int sloganOffsetY, int sloganOffsetY_B, int sloganOffsetX)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| sloganColor | int | 设置Slogan字体颜色 | 20 |
| sloganSize | int | 设置Slogan字体大小 | 10 |
| sloganOffsetY | int | 设置Slogan相对于标题栏下边缘y偏移 | 382 |
| sloganOffsetY_B | int | 设置Slogan相对于底部y偏移 | 0 |
| sloganOffsetX | int | 设置Slogan相对于屏幕左边X轴偏移量,当为0时表示居中显示 | 0 |

设置slogan文字的字体相关：setSloganViewTypeface(Typeface sloganViewTypeface)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| sloganViewTypeface | Typeface | slogan的文字的字体 | Typeface.DEFAULT |

设置隐私条款布局：setPrivacyLayout(int privacyLayoutWidth, int privacyOffsetY, int privacyOffsetY_B, int privacyOffsetX,boolean isUseNormalWebActivity)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| privacyLayoutWidth | int | 设置隐私条款宽度，隐私条款大部分为字体，只规定宽度，高度自适应 | 256 |
| privacyOffsetY | int | 设置隐私条款相对于标题栏下边缘y偏移 | 0 |
| privacyOffsetY_B | int | 设置隐私条款相对于底部y偏移	 | 18 |
| privacyOffsetX | int | 设置隐私条款对于屏幕左边X轴偏移量,当为0时表示居中显示 | 0 |
| isUseNormalWebActivity | boolean | 设置是否跳转到默认的隐私条款页面 | true |

设置隐私条款选择框相关：setPrivacyCheckBox(String unCheckedImgPath, String checkedImgPath, boolean privacyState, int privacyCheckBoxWidth, int privacyCheckBoxHeight)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| unCheckedImgPath | String | 设置未选中下按钮的图片地址| gt_one_login_unchecked |
| checkedImgPath | String | 设置选中下按钮的图片地址 | gt_one_login_checked |
| privacyState | boolean | 设置选择框是否默认选中 | true |
| privacyCheckBoxWidth | int | 选择框图片宽度 | 9 |
| privacyCheckBoxHeight | int | 选择框图片高度 | 9 |

设置隐私条款字体相关：setPrivacyClauseView(int baseClauseColor, int color, int privacyClauseTextSize)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| baseClauseColor | int | 设置隐私条款基础文字颜色	 | 0xFFA8A8A8 |
| color | int | 设置隐私条款协议文字颜色 | 0xFF3973FF |
| privacyClauseTextSize | int | 设置隐私条款字体大小 | 10 |

设置除了隐私条款其他的字体相关：setPrivacyTextView(String privacyTextViewTv1, String privacyTextViewTv2, String privacyTextViewTv3, String privacyTextViewTv4)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| privacyTextViewTv1 | String | 设置隐私条款文字1 | 登录即同意 |
| privacyTextViewTv2 | String | 设置隐私条款文字2 | 和 |
| privacyTextViewTv3 | String | 设置隐私条款文字3 | 、 |
| privacyTextViewTv4 | String | 设置隐私条款文字4 | 并使用本机号码登录 |

设置隐私栏文字的字体相关：setPrivacyClauseViewTypeface(Typeface privacyClauseBaseTextViewTypeface,Typeface privacyClauseTextViewTypeface)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| privacyClauseBaseTextViewTypeface | Typeface | 隐私栏基础的文字的字体 | Typeface.DEFAULT |
| privacyClauseTextViewTypeface | Typeface | 隐私条款的文字的字体 | Typeface.DEFAULT |

设置开发者隐私条款相关。自定义隐私条款顺序。按顺序设置，当有条款名称与条款URL一起为空时，则使用默认的运营商隐私条款。否则按照顺序排列，自定义隐私条款最多只能两个。：setPrivacyClauseText(String clauseNameOne, String clauseUrlOne, String clauseNameTwo, String clauseUrlTwo, String clauseNameThree, String clauseUrlThree)

| 参数 | 参数类型 | 说明 | 默认值 |
| --- | ------- | --- | ----- |
| clauseNameOne | String | 设置隐私条款1名称 | null |

CHANGELOG.md
Change Log
==========
Version 2.0.3 *(2020-09-17)*
----------------------------
- 修复 已知问题

Version 2.0.2 *(2020-07-17)*
----------------------------
- 优化 移除位置权限申请
- 优化 升级运营商 SDK
- 优化 其它功能优化

Version 2.0.1.6 *(2020-04-30)*
----------------------------
- 优化 OneLogin 与 OnePass 合并
- 优化 升级运营商 SDK
- 优化 其它功能优化

Version 1.0.5 *(2020-02-26)*
----------------------------
- 优化 调整缓存过期时间

Version 1.0.4 *(2020-01-09)*
----------------------------
- 优化 修复已知问题
- 优化 提供远程依赖集成

Version 1.0.3 *(2019-11-15)*
----------------------------
- 优化 修复已知问题

Version 1.0.2 *(2019-10-30)*
----------------------------
- 优化 完善逻辑

Version 1.0.1 *(2019-09-02)*
----------------------------
- 新增 自定义主题功能

app/src/main/java/com/yunpian/mobileauth/demo/ActivityAuthStyleChoose.java
package com.yunpian.mobileauth.demo;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.graphics.Color;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.cmic.sso.sdk.AuthRegisterViewConfig;
import com.cmic.sso.sdk.utils.rglistener.CustomInterface;
import com.geetest.onelogin.OneLoginHelper;
import com.geetest.onelogin.config.OneLoginThemeConfig;
import com.qipeng.yp.onelogin.QPOneLogin;
import com.qipeng.yp.onelogin.callback.AbsQPResultCallback;
import com.qipeng.yp.onelogin.callback.QPResultCallback;
import com.qipeng.yp.onelogin.callback.SimpleActivityLifecycleCallback;

import org.json.JSONException;
import org.json.JSONObject;

public class ActivityAuthStyleChoose extends AppCompatActivity {

    private View immersiveMode;
    private View dialogMode;
    private View floatMode;
    private View landscapeMode;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setStatusBarColor();
        setContentView(R.layout.activity_auth_style_choose);
        initToolbar();

        immersiveMode = findViewById(R.id.immersive_mode);
        dialogMode = findViewById(R.id.dialog_mode);
        floatMode = findViewById(R.id.float_mode);
        landscapeMode = findViewById(R.id.landscape_mode);

        landscapeMode.setBackgroundResource(R.drawable.bg_button_white);
    }

    private void readyState() {
        setEnable(immersiveMode, true);
        setEnable(dialogMode, true);
        setEnable(floatMode, true);
        setEnable(landscapeMode, true);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
    }

    private void disableState() {
        setEnable(immersiveMode, false);
        setEnable(dialogMode, false);
        setEnable(floatMode, false);
        setEnable(landscapeMode, false);
    }

    private void setEnable(View view, boolean enable) {
        if (view != null) {
            view.setAlpha(enable ? 1.0f : 0.5f);
            view.setEnabled(enable);
        }
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    private void setStatusBarColor() {
        Window window = getWindow();
        window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
        window.setStatusBarColor(getResources().getColor(R.color.choose_bg));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            window.setNavigationBarColor(getResources().getColor(R.color.choose_bg));
        }
    }

    private void initToolbar() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle("");
        setSupportActionBar(toolbar);
        final ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowHomeEnabled(true);
            actionBar.setDisplayShowTitleEnabled(true);
            actionBar.setDisplayUseLogoEnabled(false);
            actionBar.setHomeButtonEnabled(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    public void landscapeMode(View view) {
        immersiveMode(view);
    }

    public void floatMode(View view) {
        immersiveMode(view);
    }

    public void popDialogMode(View view) {
        immersiveMode(view);
    }

    public void immersiveMode(final View view) {
        disableState();

        View customView = getLayoutInflater().inflate(R.layout.custom_login, null);
        RelativeLayout.LayoutParams layoutParamsOther = new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.MATCH_PARENT, RelativeLayout.LayoutParams.WRAP_CONTENT);
        layoutParamsOther.setMargins(0, 0, 0, Utils.dip2px(this, 80));
        layoutParamsOther.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM);
        customView.setLayoutParams(layoutParamsOther);
        customView.findViewById(R.id.login_wechat_iv).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Toast.makeText(ActivityAuthStyleChoose.this, "微信登录成功", Toast.LENGTH_SHORT).show();
            }
        });
        customView.findViewById(R.id.login_weibo_iv).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Toast.makeText(ActivityAuthStyleChoose.this, "微博登录成功", Toast.LENGTH_SHORT).show();
            }
        });
        customView.findViewById(R.id.login_sms_iv).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                smsLogin();
            }
        });

        AuthRegisterViewConfig defaultAuthRegisterViewConfig = new AuthRegisterViewConfig.Builder()
                .setCustomInterface(new CustomInterface() {
                    @Override
                    public void onClick(Context context) {
                        Log.d("", "");
                    }
                })
                .setView(null)
                .setRootViewId(AuthRegisterViewConfig.RootViewId.ROOT_VIEW_ID_BODY)
                .build();
        QPOneLogin.getInstance().addOneLoginRegisterViewConfig("title_btn", defaultAuthRegisterViewConfig);

        OneLoginThemeConfig oneLoginThemeConfig;
        // 沉浸模式添加自定义入口
        if (view.getId() == R.id.immersive_mode) {
            AuthRegisterViewConfig authRegisterViewConfig = new AuthRegisterViewConfig.Builder()
                    .setCustomInterface(new CustomInterface() {
                        @Override
                        public void onClick(Context context) {
                            Log.d("", "");
                        }
                    })
                    .setView(customView)
                    .setRootViewId(AuthRegisterViewConfig.RootViewId.ROOT_VIEW_ID_BODY)
                    .build();
            QPOneLogin.getInstance().addOneLoginRegisterViewConfig("title_btn", authRegisterViewConfig);
            oneLoginThemeConfig = new OneLoginThemeConfig.Builder()
                    .setStatusBar(0xFFFFFFFF, 0xFFFFFFFF, true)
                    .setAuthNavReturnImgView("qp_ic_arrow_back", 24, 24, false, 12)
                    .setLogoImgView("ic_launcher", 88, 88, false, 100, 0, 0)
                    .setNumberView(0xFF3D424C, 24, 168, 0, 0) // 电话号码
                    .setSloganView(0x994A4A4A, 14, 216, 0, 0) // 认证服务 slogan
                    .setPrivacyClauseView(0xFF4A4A4A, 0xFF3973FF, 10)
                    .setPrivacyCheckBox("gt_one_login_unchecked", "gt_one_login_checked", false, 11, 11)
                    .setSwitchView("切换账号", 0xFF3973FF, 14, true, 249, 0, 0) // 切换账号
                    .setLogBtnLayout("bg_button", 268, 48, 260, 0, 0) // 一键登录 按钮
                    .setPrivacyLayout(300, 320, 0, 0, true) // 服务条款
                    .build();
        } else if (view.getId() == R.id.dialog_mode) {
            oneLoginThemeConfig = new OneLoginThemeConfig.Builder()
                    .setDialogTheme(true, 300, 500, 0, 0, false, false)
                    .setLogoImgView("ic_launcher", 88, 88, false, 80, 0, 0)
                    .setSwitchView("切换账号", 0xFF3973FF, 14, true, 249, 0, 0) // 切换账号
                    .setPrivacyCheckBox("gt_one_login_unchecked", "gt_one_login_checked", false, 11, 11)
                    .setPrivacyLayout(300, 0, 18, 0, true) // 服务条款
                    .build();
        } else if (view.getId() == R.id.float_mode) {
            oneLoginThemeConfig = new OneLoginThemeConfig.Builder()
                    .setDialogTheme(true, Utils.getScreenWidthForDp(this), 500, 0, 0, true, false)
                    .setLogoImgView("ic_launcher", 88, 88, false, 80, 0, 0)
                    .setSwitchView("切换账号", 0xFF3973FF, 14, true, 249, 0, 0) // 切换账号
                    .setPrivacyCheckBox("gt_one_login_unchecked", "gt_one_login_checked", false, 11, 11)
                    .setPrivacyLayout(300, 0, 18, 0, true) // 服务条款
                    .build();
        } else {
            oneLoginThemeConfig = new OneLoginThemeConfig.Builder()
                    .setStatusBar(0xFFFFFFFF, 0xFFFFFFFF, true)
                    .setAuthNavReturnImgView("qp_ic_arrow_back", 24, 24, false, 12)
                    .setLogoImgView("ic_launcher", 88, 88, false, 20, 0, 0)
                    .setNumberView(0xFF3D424C, 24, 88, 0, 0) // 电话号码
                    .setSloganView(0x994A4A4A, 11, 136, 0, 0) // 认证服务 slogan
                    .setPrivacyClauseView(0xFF4A4A4A, 0xFF3973FF, 10)
                    .setPrivacyCheckBox("gt_one_login_unchecked", "gt_one_login_checked", false, 11, 11)
                    .setSwitchView("切换账号", 0xFF3973FF, 14, true, 169, 0, 0) // 切换账号
                    .setLogBtnLayout("bg_button", 268, 48, 180, 0, 0) // 一键登录 按钮
                    .setPrivacyLayout(300, 152, 0, 0, true) // 服务条款
                    .build();
            OneLoginHelper.with().setRequestedOrientation(this, false);
        }

        QPOneLogin.getInstance().requestToken(oneLoginThemeConfig, new AbsQPResultCallback() {
            @Override
            public void onSuccess(String message) {
                readyState();
                QPOneLogin.getInstance().dismissAuthActivity();
                startActivity(new Intent(ActivityAuthStyleChoose.this, ActivityAuthResult.class));
            }

            @Override
            public void onFail(String message) {
                readyState();
                String messageResult = "OneLogin fail message = " + message;
                Log.d("kkk", "message = " + messageResult);
                try {
                    JSONObject result = new JSONObject(message);
                    int errorCode = result.optInt("errorCode");
                    if (errorCode == -20303) {
                        return;
                    }
                    if (errorCode == -20301 || errorCode == -20302) {
                        messageResult = "用户手动取消验证";
                        return;
                    }
                    if (errorCode == -20202 || errorCode == -20201) {
                        messageResult = "请确保数据流量开关已开启，然后重试";
                    }
                    if (errorCode == -20205) {
                        messageResult = "连接超时";
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }
                Toast.makeText(ActivityAuthStyleChoose.this, messageResult, Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onAuthActivityCreate(Activity activity) {
                Log.d("qipeng_log", "onAuthActivityCreate");
                if (view.getId() == R.id.landscape_mode) {
                    OneLoginHelper.with().setRequestedOrientation(activity, false);
                }
            }

            @Override
            public void onAuthWebActivityCreate(Activity activity) {
                Log.d("qipeng_log", "onAuthWebActivityCreate");
            }

            @Override
            public void onLoginButtonClick() {
                Log.d("qipeng_log", "onLoginButtonClick");
            }

            @Override
            public void onPrivacyCheckBoxClick(boolean isChecked) {
                Log.d("qipeng_log", "onPrivacyCheckBoxClick");
            }

            @Override
            public void onPrivacyClick(String name, String url) {
                Log.d("qipeng_log", "onPrivacyClick");
            }
        });
    }

    private void smsLogin() {
        if (!isNetworkConnected(this)) {
            Toast.makeText(ActivityAuthStyleChoose.this, "请确保网络连接正常", Toast.LENGTH_SHORT).show();
            return;
        }
        // 自定义短信界面
        QPOneLogin.getInstance().registerSmsAuthActivityLifecycleCallback(new SimpleActivityLifecycleCallback() {
            @Override
            public void onActivityCreated(final Activity activity, Bundle savedInstanceState) {
                // 调整虚拟导航栏颜色
                Window window = activity.getWindow();
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    window.setNavigationBarColor(Color.WHITE);
                }

                LinearLayout customView = activity.findViewById(R.id.custom_login_ll);
                customView.findViewById(R.id.login_wechat_iv).setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Toast.makeText(ActivityAuthStyleChoose.this, "微信登录成功", Toast.LENGTH_SHORT).show();
                    }
                });
                customView.findViewById(R.id.login_weibo_iv).setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Toast.makeText(ActivityAuthStyleChoose.this, "微博登录成功", Toast.LENGTH_SHORT).show();
                    }
                });
                ImageView loginBtn = customView.findViewById(R.id.login_sms_iv);
                loginBtn.setImageResource(R.drawable.ic_login_onelogin);
                loginBtn.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        activity.finish();
                    }
                });
            }
        });
        QPOneLogin.getInstance().requestSmsToken(new QPResultCallback() {
            @Override
            public void onSuccess(String message) {
                try {
                    JSONObject result = new JSONObject(message);
                    if (result.has("cid")) {
                        Toast.makeText(ActivityAuthStyleChoose.this, "验证码发送成功", Toast.LENGTH_SHORT).show();
                    }
                    if (result.optBoolean("passed")) {
                        Toast.makeText(ActivityAuthStyleChoose.this, "短信验证成功", Toast.LENGTH_SHORT).show();
                        Intent intent = new Intent(ActivityAuthStyleChoose.this, ActivityAuthResult.class);
                        intent.putExtra("title", "短信验证成功");
                        startActivity(intent);
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onFail(String message) {
                try {
                    JSONObject result = new JSONObject(message);
                    if (result.optInt("status") == 1001) {
//                        Toast.makeText(ActivityAuthStyleChoose.this, "用户手动取消验证", Toast.LENGTH_SHORT).show();
                        return;
                    }
                    String msg = result.optString("msg", "短信验证失败");
                    Toast.makeText(ActivityAuthStyleChoose.this, msg, Toast.LENGTH_SHORT).show();
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        });
    }

    public boolean isNetworkConnected(Context context) {
        if (context != null) {
            ConnectivityManager mConnectivityManager = (ConnectivityManager) context
                    .getSystemService(Context.CONNECTIVITY_SERVICE);
            NetworkInfo mNetworkInfo = mConnectivityManager.getActiveNetworkInfo();
            if (mNetworkInfo != null) {
                return mNetworkInfo.isAvailable();
            }
        }
        return false;
    }

}
